pkg/audit/manager.go

```diff
@@ -126,47 +126,91 @@ type PubsubMsg struct {
-type updateListEntry struct {
-	group             string
-	version           string
-	kind              string
-	namespace         string
-	name              string
-	msg               string
-	enforcementAction util.EnforcementAction
+// A max PriorityQueue implements heap.Interface and holds StatusViolation.
+type SVQueue []*StatusViolation
+
+func (svq SVQueue) Len() int { return len(svq) }
+
+// Implements sort.Interface based on the group, version, kind, namespace, name, message and enforcement action fields.
+// For Pop to give us the highest priority, use greater than here.
+func (svq SVQueue) Less(i, j int) bool {
+	if svq[i].Group == svq[j].Group {
+		if svq[i].Version == svq[j].Version {
+			if svq[i].Kind == svq[j].Kind {
+				if svq[i].Namespace == svq[j].Namespace {
+					if svq[i].Name == svq[j].Name {
+						if svq[i].Message == svq[j].Message {
+							return svq[i].EnforcementAction > svq[j].EnforcementAction
+						}
+						return svq[i].Message > svq[j].Message
+					}
+					return svq[i].Name > svq[j].Name
+				}
+				return svq[i].Namespace > svq[j].Namespace
+			}
+			return svq[i].Kind > svq[j].Kind
+		}
+		return svq[i].Version > svq[j].Version
+	}
+	return svq[i].Group > svq[j].Group
 }
 
-// ByGVKNNMsg implements sort.Interface based on the group, version, kind, name, namespace, and msg fields.
-type byGVKNNMsg []updateListEntry
-
-func (a byGVKNNMsg) Len() int {
-	return len(a)
+func (svq SVQueue) Swap(i, j int) {
+	svq[i], svq[j] = svq[j], svq[i]
 }
 
-func (a byGVKNNMsg) Less(i, j int) bool {
-	if a[i].group != a[j].group {
-		return a[i].group < a[j].group
-	}
-	if a[i].version != a[j].version {
-		return a[i].version < a[j].version
+func (svq *SVQueue) Push(x any) {
+	sv, ok := x.(*StatusViolation)
+	if !ok {
+		return
 	}
-	if a[i].kind != a[j].kind {
-		return a[i].kind < a[j].kind
+	*svq = append(*svq, sv)
+}
+
+func (svq *SVQueue) Pop() any {
+	old := *svq
+	n := len(old)
+	sv := old[n-1]
+	*svq = old[:n-1]
+	return sv
+}
+
+// LimitQueue implements logic to ensure priority queue len <= limit in order to provide performance guarantees on heap methods.
+type LimitQueue struct {
+	limit int
+	svq   SVQueue
+}
+
+func newLimitQueue(l int) LimitQueue {
+	lq := LimitQueue{
+		limit: l,
+		svq:   make(SVQueue, 0, l),
 	}
-	if a[i].namespace != a[j].namespace {
-		return a[i].namespace < a[j].namespace
+	heap.Init(&lq.svq)
+	return lq
+}
+
+func (lq LimitQueue) Len() int { return lq.svq.Len() }
+
+func (lq *LimitQueue) Push(x *StatusViolation) {
+	heap.Push(&lq.svq, x)
+	for lq.svq.Len() > lq.limit {
+		heap.Pop(&lq.svq)
 	}
-	if a[i].name != a[j].name {
-		return a[i].name < a[j].name
+}
+
+func (lq *LimitQueue) Pop() *StatusViolation {
+	sv, ok := heap.Pop(&lq.svq).(*StatusViolation)
+	if !ok {
+		return &StatusViolation{}
 	}
-	return a[i].msg < a[j].msg
+	return sv
 }
 
-func (a byGVKNNMsg) Swap(i, j int) {
-	a[i], a[j] = a[j], a[i]
+func (lq LimitQueue) Peek() *StatusViolation {
+	sv := lq.Pop()
```

Comment:
What is the behavior here if svq is empty? Are we adding nil?

Reply:
Good catch, I added a check for empty queue and added a check in the unit test for this behavior.